### PR TITLE
[P2] Add nnsight/NDIF backend for interventions

### PR DESCRIPTION
**Pull request by melatg0**
## Description

Adds an nnsight/NDIF backend so pyvene interventions can run remotely on NDIF (or locally through nnsight). The main point is being able to intervene on big hosted models like Llama-3.1-405B without a local GPU using the same pyvene API, just `remote=True`.

- New `IntervenableNdifModel` supporting forward, `generate`, serial interventions, and save/load, plus a torch-only helper so the code passes NDIF's server allowlist.
- `get_remote_weights()` on trainable interventions to send their weights for remote runs.
- Optional install: `pip install "pyvene[ndif]"`. Only the extra needs Python ≥ 3.10 (nnsight 0.7); core pyvene still supports 3.9.
- Adds a tutorial (`ndif_backend_101.ipynb`) and per-intervention examples.

## Testing Done

Full suite on Python 3.9 / transformers 4.57 (no regressions):

 ```
 $ python -m unittest discover -p '*TestCase.py'
    Ran 100 tests in 108s
    OK (skipped=9)
```

The 9 skipped are the opt-in remote NDIF tests. Running them against live NDIF
(Python 3.12, NDIF_REMOTE_TESTS=1, Llama-3.1-8B):
```

$ NDIF_REMOTE_TESTS=1 python -m unittest tests.integration_tests.NdifBackendTestCase
    Ran 28 tests in 52s
    OK
```

Also verified NdifBackendCorrectnessTestCase (backend matches HuggingFace and native pyvene exactly, torch.allclose 0.0), and ran ndif_backend_101.ipynb end to end (GPT-2 local + Llama-3.1-405B remote) with no errors.

## Checklist:

- [x] My PR title strictly follows the format: `[Your Priority] Your Title`
- [x] I have attached the testing log above
- [x] I provide enough comments to my code
- [x] I have changed documentations
- [x] I have added tests for my changes

### CODE REVIEW

File: pyvene/models/ndif_remote_helper.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+                src = spec.get('source_representation')
+                if src is None:
+                    src = source_activations.get(spec['key'])
+                if src is None:
+                    continue
+                src = src.to(dtype=_get_act(
+                    _get_module_output(spec['module_hook'], spec['hook_type'])
+                ).dtype)
+                output = _get_module_output(spec['module_hook'], spec['hook_type'])
+                _set_act(output, src)
```
[the remark this answers is not in the document]
Comment (melatg0) [reply pyvene/models/ndif_remote_helper.py]: Edited in most recent commit. 

The helper now reads base_loc/source_loc off the specs instead of just writing the whole activation back. Vanilla/add/subtract go through a small _apply_at helper that only touches the requested positions, and the trainable path lines the source up to the base positions and scatters just those back (which also keeps lossy transforms like the autoencoder from clobbering untargeted tokens). 

I also added a remote test that runs a single-position swap on Llama-3.1-8B and checks it matches a hand-written nnsight interchange exactly, so it now lines up with the local/native backends.

File: pyvene/models/ndif_remote_helper.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+                output = _get_module_output(spec['module_hook'], spec['hook_type'])
+                saved = _get_act(output).save()
+                model_out = output_module.output.save()
+            collected_activations[spec['key']] = saved
+
+        return {'output': model_out, 'activations': collected_activations}
+
+    # general case: gather source activations, then apply everything to base
+    source_activations = {}
+    collected_activations = {}
```
[the remark this answers is not in the document]
Comment (melatg0) [reply pyvene/models/ndif_remote_helper.py]: Also addressed in most recent commit. 

It no longer starts from an empty dict, it seeds the source activations from whatever the caller passed in via source_representations/activations_sources (handling both the bare-tensor and list-wrapped forms). Same seeding is applied across the forward, generate, and serial paths, so the intervention loops actually see the provided activations now instead of hitting src is None + no-op'ing.

**Comment by melatg0** [comment]
@codex review.

File: pyvene/models/intervenable_base.py [hunk 3]
```diff
@@ -1102,6 +1496,169 @@ def forward(
...
+                        base_act = base_output
+
+                    source_act = source_activations.get(key)
+
+                    if isinstance(intervention, TrainableIntervention) and source_act is not None:
+                        # Apply trainable intervention - gradients flow through intervention params
+                        # Get weights and apply rotation logic inline
+                        weights = intervention.get_remote_weights()
+                        rotation_matrix = weights.get('rotate_layer_weight')
+                        intervention_type = weights.get('intervention_type', 'rotated_space')
```
[the remark this answers is not in the document]
Comment (melatg0) [reply pyvene/models/intervenable_base.py]: Fixed in most recent commit.

forward_with_gradients no longer goes through get_remote_weights (which handed back detached tensors). It calls the intervention module directly via do_intervention, gathering the targeted positions first like the native backend does, so autograd stays connected to the live params. Updated test_forward_with_gradients_method to actually run loss.backward() and assert the rotation params get non-zero grads, which fails on the old detached path and passes now.

File: pyvene/models/intervenable_base.py [hunk 4]
```diff
@@ -1113,7 +1670,208 @@ def generate(
...
+        """Local generate via nnsight's generate context, covering all intervention types."""
+        activations_sources = source_representations
+        if unit_locations is None:
+            unit_locations = {}
+        if "sources->base" not in unit_locations:
+            if "base" in unit_locations:
+                unit_locations = {"sources->base": (None, unit_locations["base"])}
+            else:
+                unit_locations = {"sources->base": (None, None)}
+        unit_locations = self._broadcast_unit_locations(get_batch_size(base), unit_locations)
```
[the remark this answers is not in the document]
Comment (melatg0) [reply pyvene/models/intervenable_base.py:1731]: Addressed in most recent commit.

Both _generate_local and _generate_remote read source_loc/base_loc now instead of dropping them after the broadcast. Local gathers/scatters at the requested positions, and remote threads them into the specs so the generate helper uses the same _apply_at/_align_source/_scatter targeting as the forward path (the edit goes on the prefill pass, so the indices are valid). Also added a remote test that checks a single-position generate matches a hand-written nnsight patch token-for-token.

File: pyvene/models/ndif_remote_helper.py [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+    **kwargs
+):
+    """Run interventions on the NDIF backend. Handles every pyvene intervention
+    type, including trainable ones via their get_remote_weights()."""
+    if output_module is None:
+        output_module = model.lm_head  # models without _get_output_module
+
+    collect_specs     = [s for s in intervention_specs if s.get('is_collect')]
+    vanilla_specs     = [s for s in intervention_specs if s.get('is_vanilla') and not s.get('is_trainable')]
+    trainable_specs   = [s for s in intervention_specs if s.get('is_trainable') and not s.get('is_collect')]
```
[the remark this answers is not in the document]
Comment (melatg0) [reply pyvene/models/ndif_remote_helper.py]: Fixed in most recent commit.

The buckets key off the serialized weights now, not just is_trainable, so PCA (which has get_remote_weights but isn't a TrainableIntervention) gets routed and applied instead of falling through untouched. I also pulled the trainable math into one shared _apply_trainable_weights helper so the generate/serial paths handle PCA/autoencoder too, not just rotation. Added a remote PCA test to confirm it actually intervenes now.

**Comment by melatg0** [comment]
@codex review.
